aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -0,0 +1,67 @@
+#include <torch/library.h>
+#include <ATen/VmapTransforms.h>
+#include <ATen/ATen.h>
+
+namespace at {
+
+// NOTE: [What is a batching rule?]
+//
+// A *batching rule* implements the logic of how to call an operator on inputs
+// that have zero or more additional batch dimensions. When one does a vmap, the
+// dimension(s) being vmap'ed over get recorded as batch dimensions.
+//
+// For example, vmap(torch.add)(x, y)
+// 1. wraps `x` into batched_x = BatchedTensor(x, bdims=[(lvl=1, dim=0)];
+// 2. wraps `y` into batched_y = BatchedTensor(y, bdims=[(lvl=1, dim=0)];
+// 3. and then runs `torch.add(batched_x, batched_y)`.
+
+// NOTE: [When should I add a batching rule?]
+// When you are adding a new operator, you'll need to add a batching rule so
+// that vmap can work efficiently with said operator. If you do not, we'll attempt
+// to generate a slow fallback for the batching rule (this is not yet implemented).
+
+// NOTE: [How to write batching rules?]
+// The signature of a batching rule should look like exactly like the C++ signature
+// of its operator.
+//
+// At a high level, what a batching rule does is the following:
+// 1. Converts (logical) BatchedTensors to views on physical tensors.
```

Comment:
I think this could be easier to read for novices if you define what you mean by logical and physical first

Reply:
I'll link to the physical vs logical definition on the other source file. I agree that the current organization is not the best; ideally we'd have a "batching" or "vmap" namespace and folder that contains all of the vmap implementation details with a README that details what is going on and the right order to read the code.